proposals/0540-jsem.rst

```diff
@@ -0,0 +1,384 @@
+   which sits at the bottom of the dependency graph and must thus be compiled
+   before anything else.
+
+   a. ``cabal`` acquires one token from the semaphore and spawns one
+      ``ghc --make -jsem ⟨sem⟩`` invocation.
+   b. This invocation of ``ghc`` notices it has a lot of work to do (many modules
+      to compile from the ``Bot`` unit), so it requests more resources from the
+      semaphore. As no other processes are competing for semaphore tokens, this
+      GHC invocation obtains the remaining 7 tokens.
+   c. ``ghc`` finishes compiling the ``Bot`` unit, releasing the 7 tokens it
+      acquired.
+   d. ``cabal`` notices the ``ghc`` subprocess has terminated, and releases
+      the final (8th) token to the semaphore.
+
+3. After that, we move to compiling the middle units.
+
+   a. ``cabal`` will acquire tokens from the semaphore and spawn
+      ``ghc --make -jsem ⟨sem⟩`` invocations.
+   b. These ``ghc`` invocations won't query for more resources from the
+      semaphore, so ``cabal`` will manage running 8 concurrent ``ghc`` processes,
+      spawning new ones as previous ones terminate.
+
+4. Once all the middle units are compiled, ``cabal`` will move on to compiling
+   the top unit, which will proceed as in (2) with a single
+   ``ghc --make -jsem ⟨sem⟩`` invocation compiling 8 modules in parallel.
+
+5. Once all ``ghc`` processes have terminated, we are done, and ``cabal``
+   destroys ``⟨sem⟩``.
+
+In this situation, ``cabal`` is the jobserver: it manages the semaphore and
+spawns ``ghc`` subprocesses. The ``ghc`` subprocesses are jobclients, and they
+communicate by use of the semaphore.
+
+Effects and Interactions
+------------------------
+
+Comparison with multiple home units
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Support for multiple home units (not yet fully implemented in ``cabal``)
+would provide an alternative way to saturate the number of available capabilities.
+This is because compilation with multiple home units is achieved using a single
+GHC invocation, which thus doesn't have to worry about contention with
+other processes.
+
+In general, it would be preferred to use multiple home units when possible, as
+it is expected to be more performant than ``-jsem``:
+
+- no scheduling between different GHC invocations is necessary;
+- modules are loaded directly into the home unit graph, which avoids having
+  to load the same interface files in different GHC invocations,
+- it doesn't require the entire unit to finish compiling before compilation
+  can start on another unit that depends on it: we can begin as soon as all
+  the modules we need have been compiled.
+
+However, it's not always possible to compile everything with a single GHC
+invocation, e.g. if the build plan involves non-Haskell dependencies somewhere
+in the middle. In comparison, the ``-jsem`` functionality can fit into any build
+system that one might be using, so it supports a wider range of use cases.
```

Comment:
Based on this brief description, it sounds like the multiple home units approach is expected to be strictly better than the job-server approach for pure-Haskell build plans. That's what I would expect the vast majority of Cabal and Stack build plans to be, since neither tool has a notion of multi-language builds.

The point about polyglot builds makes sense in principle, but then the work should be motivated from that perspective, maybe with an example of how it would help bazel's `rules_haskell`. And then I'd have a number of different questions.

Reply:
I'm not familiar with `rules_haskell`, so I can't comment. For me, the real point is that the changes to build tools are very small for this approach, whereas the changes required to `cabal` for multiple home units are complex and have many subtleties.

Reply:
Otherwise, I think the point is that this feature is not relevant to `rules_haskell`, because `rules_haskell` already fully controls all parallelism because it manually invokes GHC on a per-module basis. I've made a "One-shot mode"' subsection to the "Alternatives" section to try to make that clearer.